### PR TITLE
cloud: set and handle expiration time of device state online

### PR DESCRIPTION
**Pull request by jkralik**

### CODE REVIEW

**Review by rafajpet — APPROVED** [review]

File: cloud2cloud-connector/service/devicesSubscription.go [hunk 1]
```diff
@@ -213,19 +180,16 @@ func (s *SubscriptionManager) HandleDevicesOffline(ctx context.Context, d subscr
 		authCtx := pbCQRS.AuthorizationContext{
 			DeviceId: device.ID,
 		}
-		err := s.publishCloudDeviceStatus(ctx, device.ID, authCtx, pbCQRS.CommandMetadata{
+		cmdMeta := pbCQRS.CommandMetadata{
```
Comment (ondrejtomcik) [inline cloud2cloud-connector/service/devicesSubscription.go]: Please be consistent with variable names. cmdMetadata, cmdMeta

File: coap-gateway/schema/device/status/operations.go [hunk 2]
```diff
@@ -0,0 +1,74 @@
...
+	"context"
+	"time"
+
+	"github.com/plgd-dev/cloud/resource-aggregate/cqrs/utils"
+	"github.com/plgd-dev/cloud/resource-aggregate/pb"
+	"github.com/plgd-dev/go-coap/v2/message"
+	"github.com/plgd-dev/kit/codec/cbor"
+)
+
+// Publish publish device cloud state resource.
```
Comment (ondrejtomcik) [inline coap-gateway/schema/device/status/operations.go]: publish publish ?

File: coap-gateway/schema/device/status/operations.go [hunk 3]
```diff
@@ -0,0 +1,74 @@
...
+func Publish(ctx context.Context, client pb.ResourceAggregateClient, deviceID string, cmdMeta *pb.CommandMetadata, authCtx *pb.AuthorizationContext) error {
+	_, err := client.PublishResource(ctx, &pb.PublishResourceRequest{
```
Comment (ondrejtomcik) [inline coap-gateway/schema/device/status/operations.go:15]: This is resource publish, it doesn't contain device status - just refering to documentation of this method.

File: coap-gateway/schema/device/status/operations.go [hunk 4]
```diff
@@ -0,0 +1,74 @@
...
+
+	return err
+}
+
+func update(ctx context.Context, client pb.ResourceAggregateClient, deviceID string, online bool, expires time.Time, cmdMeta *pb.CommandMetadata, authCtx *pb.AuthorizationContext) error {
+	data, err := cbor.Encode(Status{
+		ResourceTypes: ResourceTypes,
+		Interfaces:    Interfaces,
+		Online:        online,
+		Expires:       expires.Unix(),
```
Comment (ondrejtomcik) [inline coap-gateway/schema/device/status/operations.go]: What does it mean - Expires. Expires in ? Valid until ?

File: coap-gateway/schema/device/status/status.go [hunk 5]
```diff
@@ -0,0 +1,30 @@
...
+
+const ResourceType string = "x.cloud.device.status"
+const Title string = "Device cloud status"
+
+var Interfaces = []string{"oic.if.baseline"}
+var ResourceTypes = []string{ResourceType}
+
+const Href = "/plgd.dev/s"
+
+// Status is resource published by OCF Cloud.
```
Comment (ondrejtomcik) [inline coap-gateway/schema/device/status/status.go]: published and maintained by plgd cloud.

File: coap-gateway/schema/device/status/status.go [hunk 6]
```diff
@@ -0,0 +1,30 @@
...
+// - signup: resource published
+// - signin: content changed -> online true and set expires to unix timestap when online state expires (0 means never)
+// - signout/close connection: content changed -> online false
+type Status struct {
+	ResourceTypes []string `json:"rt"`
+	Interfaces    []string `json:"if"`
+	Online        bool     `json:"online"`
+	Expires       int64    `json:"expires"`
```
Comment (ondrejtomcik) [inline coap-gateway/schema/device/status/status.go]: validUntil?

File: coap-gateway/schema/device/status/status.go [hunk 7]
```diff
@@ -0,0 +1,30 @@
...
+}
+
+// IsOnline evaluate online state
+func (s Status) IsOnline() bool {
+	if !s.Online || s.Expires <= 0 {
```
Comment (ondrejtomcik) [inline coap-gateway/schema/device/status/status.go]: what is the case for s.Expires <= 0?

File: coap-gateway/schema/device/status/status.go [hunk 8]
```diff
@@ -0,0 +1,30 @@
...
+		return s.Online
+	}
+	return time.Now().Before(time.Unix(s.Expires, 0))
```
Comment (ondrejtomcik) [inline coap-gateway/schema/device/status/status.go]: Do you have a test for this? Can it be negative?

File: coap-gateway/schema/device/status/operations.go (hunk 4 above)
Comment (jkralik) [reply coap-gateway/schema/device/status/operations.go]: Valid until

Comment (ondrejtomcik) [reply coap-gateway/schema/device/status/operations.go]: Can you rename it everywhere please?

File: coap-gateway/schema/device/status/status.go (hunk 7 above)
Comment (jkralik) [reply coap-gateway/schema/device/status/status.go]: Infinite

File: coap-gateway/schema/device/status/status.go (hunk 8 above)
Comment (jkralik) [reply coap-gateway/schema/device/status/status.go]: It cannot be negative .

File: coap-gateway/schema/device/status/operations.go (hunk 2 above)
Comment (jkralik) [reply coap-gateway/schema/device/status/operations.go]: Publish publishes the device cloud state resource.

**Review by ondrejtomcik — APPROVED** [review]

**Pull request merged**